torch/csrc/jit/codegen/cuda/compute_at_map.cpp

```diff
@@ -931,130 +869,233 @@ IterDomain* ComputeAtMap::computeConcreteId(
+        }
+      }
+    }
+
+    // Remove all sets in rfactor history from all_exact_sets_covered by
+    // effectively doing an inplace copy_if
+    VectorOfUniqueEntries<std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>>
+        tmp_all_exact_sets_covered;
+    std::swap(tmp_all_exact_sets_covered, all_exact_sets_covered);
+    for (auto entry : tmp_all_exact_sets_covered) {
+      if (produces_rfactor_dom.has(entry)) {
+        continue;
+      }
+      all_exact_sets_covered.pushBack(entry);
+    }
+  }
 
-    if (mode == IdMappingMode::LOOP) {
-      std::transform(
-          root_ids.begin(),
-          root_ids.end(),
-          std::inserter(root_ids_of_all_ids, root_ids_of_all_ids.end()),
-          [&](const auto root_id) {
-            auto exact_concrete_id =
-                getConcreteMappedID(root_id, IdMappingMode::EXACT);
-            root_id_to_maybe_concrete_ids[exact_concrete_id].push_back(
-                maybe_concrete_id);
-            return exact_concrete_id;
-          });
+  VectorOfUniqueEntries<std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>>
+      input_ids;
+
+  {
+    // Remove any concrete id that's not still in all_exact_sets_covered,
+    // basically copy_if
+    decltype(maybe_concrete_ids) tmp_maybe_concrete_ids;
+    std::swap(maybe_concrete_ids, tmp_maybe_concrete_ids);
+    for (auto entry : tmp_maybe_concrete_ids) {
+      if (all_exact_sets_covered.has(
+              disjointSetOf(entry, IdMappingMode::EXACT))) {
+        maybe_concrete_ids.pushBack(entry);
+      }
     }
+  }
+
+  TORCH_INTERNAL_ASSERT(
+      maybe_concrete_ids.vector().size() > 0,
+      "No potential concrete_id's found for disjoint set ",
+      disjoint_set_shared_ptr->toString());
+
+  if (maybe_concrete_ids.vector().size() == 1) {
+    return maybe_concrete_ids.vector().front();
+  }
+
+  // The concrete_id should have the most roots it can trace back to that are
+  // iter domains, (non-broadcast/non-reduction). We don't trace back through
+  // view operations, so the one with the most iter root domains is the concrete
+  // ID.
+  IterDomain* concrete_id = nullptr;
+  int max_iter_root_count = 0;
+  int max_bcast_root_count = 0;
+
+  for (auto maybe_concrete_id : maybe_concrete_ids.vector()) {
+    auto concrete_id_root_sets = getInputDisjointSetsOf(maybe_concrete_id);
 
     int bcast_root_count = std::count_if(
-        root_ids.begin(), root_ids.end(), [&](IterDomain* root_id) {
-          return root_id->isBroadcast()
-              // TODO: This shouldn't have a negative impact, but (emperically)
-              // might not be necessary
-              || trivial_reduction_info_.isDerived(root_id);
+        concrete_id_root_sets.vector().begin(),
+        concrete_id_root_sets.vector().end(),
+        [&](std::shared_ptr<VectorOfUniqueEntries<IterDomain*>> set) {
+          return set->vector()[0]->isBroadcast() ||
+              set->vector()[0]->isTrivialReduction() ||
```

Comment:
Found some examples. I see why `trivial_reduction_info_` isn't sufficient. As we're moving away from trivial reductions, this may become a non issue soon.